### PR TITLE
Update budget recommendation API to support multiple countries

### PR DESCRIPTION
**Pull request by ianlin**
### Changes proposed in this Pull Request:

Part of #1248. This PR updates the API `ads/campaigns/budget-recommendation/:country-code` to `ads/campaigns/budget-recommendation?country_codes=` in order to support multiple countries' budget recommendation.

### Screenshots

Screenshot 2022-02-18 at 17 52 53

### Detailed test instructions:

Since the frontend hasn't integrated with this new API to show the budget recommendation for multiple countries, one can test this PR through any HTTP clients like Postman.

#### Request

```
GET https://testdomain/wp-json/wc/gla/ads/campaigns/budget-recommendation?country_codes=TW,GB,US
```

#### Response

```json
{
    "currency": "EUR",
    "recommendations": [
        {
            "country": "US",
            "daily_budget_low": 10,
            "daily_budget_high": 25
        },
        {
            "country": "GB",
            "daily_budget_low": 5,
            "daily_budget_high": 20
        },
        {
            "country": "IE",
            "daily_budget_low": 5,
            "daily_budget_high": 15
        }
    ]
}
```

### Additional details

The type of the budget in `daily_budget_low` and `daily_budget_high` was of `string`, I changed the type to `number` in order to align with the type of the response data `amount` when calling `ads/campaigns` API.

### Changelog entry

> Update - Budget recommendation for multiple countries

### CODE REVIEW

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php [hunk 1]
```diff
@@ -113,24 +142,24 @@ protected function get_schema_properties(): array {
...
-				'type'              => 'string',
-				'description'       => __( 'Country code in ISO 3166-1 alpha-2 format.', 'google-listings-and-ads' ),
+			'country_codes'     => [
+				'type'              => [ 'string' ],
+				'description'       => __( 'An array of Country codes in ISO 3166-1 alpha-2 format.', 'google-listings-and-ads' ),
 				'context'           => [ 'view' ],
 				'sanitize_callback' => $this->get_country_code_sanitize_callback(),
 				'validate_callback' => $this->get_country_code_validate_callback(),
 				'required'          => true,
 			],
```
Comment (mikkamp) [inline src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: I think we should try and use a valid type for this, since it's an array of strings you would define that as shown here: https://developer.wordpress.org/rest-api/extending-the-rest-api/schema/#arrays

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php [hunk 2]
```diff
@@ -68,31 +68,60 @@ public function register_routes(): void {
 	 */
 	protected function get_budget_recommendation_callback(): callable {
 		return function( Request $request ) {
-			$country        = strtoupper( $request->get_param( 'country_code' ) );
-			$currency       = $this->middleware->get_ads_currency();
-			$recommendation = $this
+			$country_codes = $request->get_query_params()['country_codes'] ?? '';
```
Comment (mikkamp) [inline src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: If we are getting one specific parameter I think it's better to use `get_param( 'country_codes' )`. We have already registered this as a GET route so it will always have to be a query parameter.

WordPress has built in functionality for validating the Rest API parameters, so we don't need to default to empty here. Although that's not happening in this controller, because when we register the route we don't provide the `args`, you can see an example of how that's done [here](https://github.com/woocommerce/google-listings-and-ads/blob/1.11.1/src/API/Site/Controllers/Ads/CampaignController.php#L79)

In a lot of our controllers we use the same schema both for validating the parameters and also for sanitizing the response. Either one could work here, but some of the schema needs to be adjusted to reflect how it's validated and how it's sanitized.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php [hunk 3]
```diff
@@ -68,31 +68,60 @@ public function register_routes(): void {
...
+			if ( ! $country_codes || ! $currency ) {
+				return new Response(
+					[
+						'message'       => __( 'Invalid country_codes/currency combination', 'google-listings-and-ads' ),
+						'currency'      => $currency,
+						'country_codes' => $country_codes,
+					],
+					400
+				);
+			}
```
Comment (mikkamp) [inline src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: I think we should leave this to just validate the currency since that's an internal value. We should validate the country code using the built in Rest API validation.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php [hunk 4]
```diff
@@ -68,31 +68,60 @@ public function register_routes(): void {
...
+
+			$country_codes = strtoupper( $country_codes );
+
+			$recommendations = $this
 				->budget_recommendation_query
-				->where( 'country', $country )
+				->where( 'country', explode( ',', $country_codes ), 'IN' )
```
Comment (mikkamp) [inline src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: I think it would be better to pass country codes as an array instead of a comma separated list.  Search for the function `get_collection_params` to see some examples of how we pass parameters in the query. If we setup the validation the same then it will be handled for us and we don't need to do an explode here.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php [hunk 5]
```diff
@@ -68,31 +68,60 @@ public function register_routes(): void {
...
+							return (float) $recommendation['daily_budget_low'];
+						},
+						$recommendations
+					),
+					'daily_budget_high' => array_map(
+						function ( $recommendation ) {
+							return (float) $recommendation['daily_budget_high'];
+						},
+						$recommendations
+					),
```
Comment (mikkamp) [inline src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: I think this structure isn't as clear as it should be because it results in the following:
```json
{
	"currency": "EUR",
	"country_codes": [
		"US",
		"GB",
		"IE"
	],
	"daily_budget_low": [
		10,
		5,
		5
	],
	"daily_budget_high": [
		25,
		20,
		15
	]
}
```

Which means we have to assume that all arrays are in the same order and we need to use the country_codes array as the index for the other two. As far as I recall we've stuck to a pattern like this for most of our controllers which return multiple results:
```json
{
	"currency": "EUR",
	"recommendations": [
		{
			"country": "US",
			"daily_budget_low": 10,
			"daily_budget_high": 25
		},
		{
			"country": "GB",
			"daily_budget_low": 5,
			"daily_budget_high": 20
		},
		{
			"country": "IE",
			"daily_budget_low": 5,
			"daily_budget_high": 15
		}
	]
}
```

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php [hunk 6]
```diff
@@ -113,24 +142,24 @@ protected function get_schema_properties(): array {
...
 				'required'          => true,
 			],
 			'daily_budget_low'  => [
-				'type'              => 'number',
-				'description'       => __( 'The lower limit for the recommended budget.', 'google-listings-and-ads' ),
+				'type'              => [ 'number' ],
+				'description'       => __( 'An array of the lower limit for the recommended budget.', 'google-listings-and-ads' ),
 				'context'           => [ 'view' ],
 				'validate_callback' => 'rest_validate_request_arg',
 				'required'          => true,
```
Comment (mikkamp) [inline src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: I realize these were already here, but if we aren't validating then these won't do anything. And if we do turn on validation, this will fail because `required = true` means it needs to always be passed as a parameter. If we are going to use the same schema for parameters we could set `readonly = true` but I think it would be better to have different schema for the parameters.

File: tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+			'daily_budget_high' => [930, 625, 255, 320],
+		];
+
+		$this->middleware->expects( $this->once() )
+			->method( 'get_ads_currency' )
+			->willReturn( 'TWD' );
+
+		$this->budget_recommendation_query->expects( $this->exactly(2) )
+			->method( 'where' )
+			->willReturn( $this->budget_recommendation_query );
```
Comment (mikkamp) [inline tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php]: I haven't seen any good examples of mocking chained functions, but I think it would look a little cleaner if we just move this to the setUp function and have it always return itself:

```php
$this->budget_recommendation_query->method( 'where' )->willReturn( $this->budget_recommendation_query );
```

File: tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+ */
+class BudgetRecommendationControllerTest extends RESTControllerUnitTest {
+
+	protected const ROUTE_BUDGET_RECOMMENDATION  = '/wc/gla/ads/campaigns/budget-recommendation';
+
+	public function setUp() {
+		parent::setUp();
+
+		$this->budget_recommendation_query = $this->createMock( BudgetRecommendationQuery::class );
+		$this->iso_provider = $this->createMock( ISO3166DataProvider::class );
```
Comment (mikkamp) [inline tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php:33]: If we get the validation implemented then we can change the test `test_get_budget_recommendation_with_nonexistent_country_code` to have this provider return false instead of an empty result. Otherwise we aren't using this mocked class.

**Review by mikkamp — COMMENTED** [review]
Thanks for the PR, it looks good and it returns the expected results.

However I left some comments on things I think we can refactor to stick a bit closer to the WP REST API standards.

**Comment by ianlin** [comment]
Thanks @mikkamp, they are very valuable comments! I'm adjusting the code based on your suggestions above and I will let  you know when I finish it.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php (hunk 5 above)
Comment (ianlin) [reply src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Good point, the new format is much clearer. Fixed in 07669b2.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php (hunk 1 above)
Comment (ianlin) [reply src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Thanks for the info. Updated in eb0425b to use a new schema for query parameters.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php (hunk 2 above)
Comment (ianlin) [reply src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Thanks for the info. Also updated in eb0425b to use a new schema for query parameters, and to use `get_params` instead of `get_query_params` to get the `country_codes` parameter.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php (hunk 3 above)
Comment (ianlin) [reply src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Thanks for the info. Also updated in eb0425b to use a new schema for validating and sanitising query parameters, and remove the `country_codes` validation in the controller.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php (hunk 4 above)
Comment (ianlin) [reply src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Thanks for the info. Also updated in eb0425b to get the `country_codes` array instead of a comma separated string.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php (hunk 6 above)
Comment (ianlin) [reply src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Thanks for the info. Also updated in eb0425b to use a new schema for validating the parameters.

File: tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php (hunk 7 above)
Comment (ianlin) [reply tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php]: Indeed, this is much cleaner. Updated in 4aff2bb to move the mocking `where` clause to the setUp function.

File: tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php (hunk 8 above)
Comment (ianlin) [reply tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php:33]: Thanks for spotting this! Updated in eb0425b to have the iso provider throws an exception for testing.

**Comment by ianlin** [comment]
Hey @mikkamp, I've done updating the code based on your suggestions, please review it again when you're available, thanks!

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php [hunk 9]
```diff
@@ -52,47 +52,89 @@ public function __construct( RESTServer $rest_server, BudgetRecommendationQuery
...
+
 	/**
 	 * @return callable
 	 */
 	protected function get_budget_recommendation_callback(): callable {
 		return function( Request $request ) {
-			$country        = strtoupper( $request->get_param( 'country_code' ) );
-			$currency       = $this->middleware->get_ads_currency();
-			$recommendation = $this
+			$country_codes = $request->get_params()['country_codes'];
```
Comment (mikkamp) [inline src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Why not just `get_param`? One of the warnings that appears is `Undefined index: country_codes` if we send a request without country_codes (see comment on required collection parameter).

```suggestion
			$country_codes = $request->get_param( 'country_codes' );
```

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php [hunk 10]
```diff
@@ -52,47 +52,89 @@ public function __construct( RESTServer $rest_server, BudgetRecommendationQuery
...
+		return [
+			'context'       => $this->get_context_param( [ 'default' => 'view' ] ),
+			'country_codes' => [
+				'type'              => 'array',
+				'sanitize_callback' => $this->get_country_code_sanitize_callback(),
+				'validate_callback' => $this->get_country_code_validate_callback(),
+				'items'             => [
+					'type' => 'string',
+				],
+			],
```
Comment (mikkamp) [inline src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Why not set this to `required => true`? I get several PHP warnings when I call the request without `country_codes`, setting it to required will prevent that from happening and show a standard error.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php [hunk 11]
```diff
@@ -52,47 +52,89 @@ public function __construct( RESTServer $rest_server, BudgetRecommendationQuery
...
-			$country        = strtoupper( $request->get_param( 'country_code' ) );
-			$currency       = $this->middleware->get_ads_currency();
-			$recommendation = $this
+			$country_codes = $request->get_params()['country_codes'];
+			$currency      = $this->middleware->get_ads_currency();
+
+			if ( ! $currency ) {
+				return new Response(
+					[
+						'message'       => __( 'Invalid currency', 'google-listings-and-ads' ),
```
Comment (mikkamp) [inline src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: I'm wondering how helpful this error is. Looking at the function `get_ads_currency` it seems it will always return something and if not it will default to the WooCommerce currency. So checking for a "false" value of currency doesn't seem like it would happen. And if it would (because in theory it's still possible to have it return a false value), then we should say that the error is something along the lines of "No currency available for the Ads account", we don't know whether it's a valid/invalid currency.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php [hunk 12]
```diff
@@ -52,47 +52,89 @@ public function __construct( RESTServer $rest_server, BudgetRecommendationQuery
...
 				return new Response(
 					[
-						'message'      => __( 'Invalid country/currency combination', 'google-listings-and-ads' ),
-						'currency'     => $currency,
-						'country_code' => $country,
+						'message'       => __( 'Cannot find any budget recommendations', 'google-listings-and-ads' ),
+						'currency'      => $currency,
+						'country_codes' => $country_codes,
 					],
 					400
```
Comment (mikkamp) [inline src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: If we specifically know it's going to be a "not found" error then it's probably better to use 404 as the status code.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php [hunk 13]
```diff
@@ -106,34 +148,38 @@ protected function get_budget_recommendation_callback(): callable {
...
+							'type'              => 'string',
+							'description'       => __( 'Country code in ISO 3166-1 alpha-2 format.', 'google-listings-and-ads' ),
+							'context'           => [ 'view' ],
+							'sanitize_callback' => $this->get_country_code_sanitize_callback(),
+							'validate_callback' => $this->get_country_code_validate_callback(),
+							'required'          => true,
+						],
+						'daily_budget_low'  => [
+							'type'        => 'number',
+							'description' => __( 'Daily budget lower bound ', 'google-listings-and-ads' ),
```
Comment (mikkamp) [inline src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Since it's the field description we can use a full sentence here and end it with a full stop.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php [hunk 14]
```diff
@@ -106,34 +148,38 @@ protected function get_budget_recommendation_callback(): callable {
...
+				'items' => [
+					'type'       => 'object',
+					'properties' => [
+						'country'           => [
+							'type'              => 'string',
+							'description'       => __( 'Country code in ISO 3166-1 alpha-2 format.', 'google-listings-and-ads' ),
+							'context'           => [ 'view' ],
+							'sanitize_callback' => $this->get_country_code_sanitize_callback(),
+							'validate_callback' => $this->get_country_code_validate_callback(),
+							'required'          => true,
```
Comment (mikkamp) [inline src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: The sanitation and validation is important when the schema is used for editing/creating, but here it doesn't really do much, I think just having them in the collection_params function is enough. Also required indicates which fields should be present in a request, so I think we can remove those from all the fields here.

File: tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+					'daily_budget_high' => 320.0,
+				]
+			],
+		];
+
+		$this->middleware->expects( $this->once() )
+			->method( 'get_ads_currency' )
+			->willReturn( 'TWD' );
+
+		$this->budget_recommendation_query->expects( $this->exactly(2) )->method( 'where' );
```
Comment (mikkamp) [inline tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php]: Is this assertion really useful? We aren't testing the query since it's returning mocked results, so in my opinion it would make sense to do assertions on what we are expecting as the controller response instead.
In my opinion it would only be helpful to check if we are confirming that the the right parameters were sent to the `where` function for each call. But that might be a bit out of scope for this test and better belong in a separate test where the query class is tested.

File: tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+			->method( 'get_results' )
+			->willReturn( $budget_recommendation_data );
+
+		$response = $this->do_request( self::ROUTE_BUDGET_RECOMMENDATION, 'GET', $budget_recommendation_params );
+
+		$this->assertSame( $expected_response_data, $response->get_data() );
+		$this->assertEquals( 200, $response->get_status() );
+	}
+
+	public function test_get_budget_recommendation_without_query_parameters() {
```
Comment (mikkamp) [inline tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php:109]: It looks more like we are testing with an empty string instead of without query parameters. If I test the same in postman:

`GET https://domain.test/wp-json/wc/gla/ads/campaigns/budget-recommendation?country_codes=`
I get the error 'Not a valid alpha2 key'

I guess that's because the iso provider is mocked so it doesn't throw the validation error in the test. But it's probably better to focus on the error we expect without sending any parameters (after setting country_codes to required).

If we want to specifically test for the response `Cannot find any budget recommendations`, I think we should create a separate test where we do pass the countries in the request but get an empty response from the query. A scenario like that could happen if the recommendations DB import failed.

**Review by mikkamp — COMMENTED** [review]
Thanks for all the changes, those responses look a lot cleaner that way.

I left some additional comments, it doesn't seem the unit tests are totally lined up with what we actually receive in invalid requests, so it's probably better to make that a bit more similar so we can catch any potential errors with the unit tests.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php (hunk 9 above)
Comment (ianlin) [reply src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Ah my bad, I replaced it from `get_query_params` and kept the `params` so it became `get_params` 😓. I should check the log messages to see if there are any warnings. Thanks for pointing this out!

Updated in 37e5212.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php (hunk 10 above)
Comment (ianlin) [reply src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Thanks! Also updated in 37e5212.

File: tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php (hunk 16 above)
Comment (ianlin) [reply tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php:109]: > It looks more like we are testing with an empty string instead of without query parameters.

Thanks for pointing this out! I updated it in 37e5212, I think this is because I did not set `required => true` in the collection params schema so the built-in validator didn't respond the error 400, also as you said the iso provider was mocked so it didn't respond the correct "invalid alpha2 key" error in the unit test.

> If we want to specifically test for the response Cannot find any budget recommendations, I think we should create a separate test where we do pass the countries in the request but get an empty response from the query. A scenario like that could happen if the recommendations DB import failed.

For this I updated in 57b73b9.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php (hunk 11 above)
Comment (ianlin) [reply src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Agree. Updated in eb200bb, thanks!

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php (hunk 12 above)
Comment (ianlin) [reply src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Absolutely right! Updated in 57b73b9. I should take more care on it, this was one of my interview questions for the candidates in my previous job. 😅

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php (hunk 13 above)
Comment (ianlin) [reply src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Thanks and noted! Updated in 7562c39.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php (hunk 14 above)
Comment (ianlin) [reply src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Thanks and noted again! Updated in 5a59ebd.

File: tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php (hunk 15 above)
Comment (ianlin) [reply tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php]: Agree. Updated in a9ce1e2.

**Comment by ianlin** [comment]
Hello @mikkamp, thanks for all the comments you gave, they are really useful. I've updated the code based on your suggestions, please have it a look again when you're available.

File: src/API/Site/Controllers/Ads/BudgetRecommendationController.php [hunk 17]
```diff
@@ -52,47 +52,90 @@ public function __construct( RESTServer $rest_server, BudgetRecommendationQuery
...
+					404
 				);
 			}
 
+			$returned_recommendations = array_map(
+				function ( $recommendation ) {
+					return [
+						'country'           => $recommendation['country'],
+						'daily_budget_low'  => (float) $recommendation['daily_budget_low'],
+						'daily_budget_high' => (float) $recommendation['daily_budget_high'],
```
Comment (mikkamp) [inline src/API/Site/Controllers/Ads/BudgetRecommendationController.php]: Do we need these to be cast to floats? In the database the columns are defined as `int(11)` so seems like casting to an int would do here.

File: tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+			]
+		];
+
+		$expected_response_data = [
+			'currency'        => 'TWD',
+			'recommendations' => [
+				[
+					'country'           => 'US',
+					'daily_budget_low'  => 330.0,
+					'daily_budget_high' => 930.0,
```
Comment (mikkamp) [inline tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php]: Even though we cast to float I seem to always be getting these values as 330 or 930, but never with the decimal added. Can we drop the decimals since it doesn't get returned that way in a response? It seems they are only added in the test since we are comparing with assertSame (which is strict about comparing the type), but if we cast them to int that's no longer needed.
This is what I'm seeing in the response even if they are floats, so I'd expect us to be testing with similar values:
```json
{
	"country": "US",
	"daily_budget_low": 10,
	"daily_budget_high": 25
}
```

**Review by mikkamp — APPROVED** [review]
Thanks for all those extra changes, it's working well, and I'm getting the expected results. So I'll go ahead and mark it as approved.

Tested:
- unit tests
- response with invalid parameters
- response with multiple country code
- response with invalid country code

I just left two small comments on the numbers being float values, doesn't seem like that's necessary and it's a bit confusing in the test since we aren't actually getting them with the decimals in the response.
